tensorflow-frameworks/tensorflow-data/src/test/java/org/tensorflow/data/DatasetOpTester.java

```diff
@@ -0,0 +1,283 @@
+package org.tensorflow.data;
+
+import org.junit.Test;
+import org.tensorflow.*;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.core.Constant;
+import org.tensorflow.op.data.*;
+import org.tensorflow.tools.Shape;
+import org.tensorflow.types.TInt32;
+import org.tensorflow.utils.Tuple2;
+
+import java.nio.IntBuffer;
+import java.util.Arrays;
+import java.util.List;
+
+public class DatasetOpTester {
+
+  @Test
+  public void testEagerBatching() {
+    try (EagerSession session = EagerSession.create()) {
+      Ops tf = Ops.create(session);
+
+      Constant<TInt32> X = tf.val(
+          new int[][]{
+              {1, 2, 3},
+              {4, 5, 6},
+              {7, 8, 9},
+              {10, 11, 12}
+          }
+      );
+
+      Constant<TInt32> y = tf.val(
+          new int[][]{
+              {1},
+              {4},
+              {7},
+              {10}
+          }
+      );
+
+
+      List<Operand<?>> tensors = Arrays.asList(X, y);
+
+      // // Try running TensorDataset
+      List<DataType<?>> outputTypes = Arrays.asList(TInt32.DTYPE, TInt32.DTYPE);
+      List<Shape> outputShapes = Arrays.asList(
+          Shape.of(3),
+          Shape.of(1));
+
+      TensorSliceDataset tensorDataset = TensorSliceDataset.create(tf.scope(), tensors, outputShapes);
+
+      BatchDataset batchDataset = BatchDataset.create(
+          tf.scope(),
+          tensorDataset,
+          tf.val(2L),
+          tf.val(true),
+          outputTypes,
+          Arrays.asList(
+              Shape.of(2, 3),
+              Shape.of(2, 1))
+      );
+
+
+      AnonymousIterator anonymousIter = AnonymousIterator.create(tf.scope(), outputTypes, Arrays.asList(
+          Shape.of(2, 3),
+          Shape.of(2, 1)));
+
+      MakeIterator makeIterator = tf.data.makeIterator(batchDataset, anonymousIter.handle());
+
+      while (true) {
+        try {
+          IteratorGetNext getNext = tf.data.iteratorGetNext(anonymousIter.handle(), outputTypes, outputShapes);
+          List<Output<?>> outputs = getNext.components();
+          System.out.println("BATCH: ");
+          printIntTensor(outputs.get(0).tensor());
+          printIntTensor(outputs.get(1).tensor());
+          System.out.println();
+        } catch (IndexOutOfBoundsException e) {
```

Comment:
again, any way to avoid catching an exception as part of the normal process? I think I already asked you this question before and you told me that is how Python is doing it... but still, can we do better? :)

Reply:
I can't think of a way to improve this, without restricting what can be done within the loop (e.g. arbitrary session runs / fetches / etc. that depend on the loop iteration).

Reply:
Will resolve this since we have other threads on this topic.

Reply:
Will resolve this thread as we have others on this topic.